### PR TITLE
fix: Inserting Slots within Slots from the Marketplace causes infinite loop and crash

### LINKED ISSUE #4111: Inserting Slots within Slots from the Marketplace causes infinite loop and crash
1. Add Marketplace item with Slot such as Ani Motion > Code
2. Click into the inserted Slot
3. Add the same Slot again from the Marketplace
4. Notice Builder freezes and will crash if you insert more

Seems to only be repro with Marketplace.

https://github.com/user-attachments/assets/fcbcd278-eda5-4cae-a979-a9dc6466aef4

### PR DESCRIPTION
**Pull request by istarkov**
## Description

closes #4111

- [x] - Fixes client side
- [x] - Fixes server side

https://github.com/user-attachments/assets/8bb6ffdc-2e78-45eb-b98b-e697daee822a

## Steps for reproduction

Now seems like impossible to save broken project. Tested locally on broken project.

## Code Review

- [ ] hi @kof, I need you to do
  - conceptual review (architecture, feature-correctness)
  - detailed review (read every line)
  - test it on preview

## Before requesting a review

- [ ] made a self-review
- [ ] added inline comments where things may be not obvious (the "why", not "what")

## Before merging

- [ ] tested locally and on preview environment (preview dev login: 5de6)
- [ ] updated [test cases](https://github.com/webstudio-is/webstudio/blob/main/apps/builder/docs/test-cases.md) document
- [ ] added tests
- [ ] if any new env variables are added, added them to `.env` file

### CODE REVIEW

File: packages/project-build/src/db/build.ts [hunk 1]
```diff
@@ -25,17 +25,66 @@ import { parseDeployment } from "./deployment";
 import { serializePages } from "./pages";
 import { createDefaultPages } from "../shared/pages-utils";
 import type { MarketplaceProduct } from "../shared//marketplace";
+import { findCycles } from "../shared/graph-utils";
 
 const parseCompactData = <Item>(serialized: string) =>
   JSON.parse(serialized) as Item[];
 
+const parseCompactInstanceData = (serialized: string) => {
```
Comment (kof) [inline packages/project-build/src/db/build.ts:33]: sounds like this could use a test

Comment (istarkov) [reply packages/project-build/src/db/build.ts:33]: Done

File: packages/project-build/src/shared/graph-utils.test.ts [hunk 2]
```diff
@@ -0,0 +1,148 @@
+// graph-utils.test.ts
```
Comment (TrySound) [inline packages/project-build/src/shared/graph-utils.test.ts:1]: why have a filename in comment?

File: packages/project-build/src/shared/graph-utils.ts [hunk 3]
```diff
@@ -0,0 +1,93 @@
+import type { Instance } from "@webstudio-is/sdk";
+
+type InstanceId = Instance["id"];
+
+// Depth-First Search (DFS) algorithm to find cycles in a directed graph
+export const findCycles = (
+  graph: Iterable<Pick<Instance, "id" | "children">>
+): InstanceId[][] => {
+  const adjacencyList: Record<InstanceId, InstanceId[]> = {};
```
Comment (TrySound) [inline packages/project-build/src/shared/graph-utils.ts:9]: Maybe use map for it?

File: packages/project-build/src/shared/graph-utils.test.ts (hunk 2 above)
Comment (istarkov) [reply packages/project-build/src/shared/graph-utils.test.ts]: Auto generated test, 99% is written with vscode chat gpt

File: packages/project-build/src/shared/graph-utils.ts (hunk 3 above)
Comment (istarkov) [reply packages/project-build/src/shared/graph-utils.ts:9]: Not needed, we are just iterating over it

**Review by TrySound — APPROVED** [review]

File: packages/project-build/src/shared/graph-utils.ts [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+    }
+
+    if (visited.has(nodeId)) {
+      return;
+    }
+
+    visited.add(nodeId);
+    path.push(nodeId);
+
+    for (const childId of adjacencyList[nodeId] || []) {
```
Comment (TrySound) [inline packages/project-build/src/shared/graph-utils.ts:36]: Map is more type safe here

File: packages/project-build/src/shared/graph-utils.ts [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+      .map((child) => child.value);
+  }
+
+  const visited = new Set<string>();
+  const path: InstanceId[] = [];
+  const cycles: InstanceId[][] = [];
+
+  const dfs = (nodeId: string): void => {
+    if (path.includes(nodeId)) {
+      const cycleStart = path.indexOf(nodeId);
```
Comment (TrySound) [inline packages/project-build/src/shared/graph-utils.ts:24]: could be single path.indexOf instead of both includes and insteadOf

Comment (istarkov) [reply packages/project-build/src/shared/graph-utils.ts:24]: Usually in this algorithms here is not inlined code but some `normalizePath` useless for our purposes.

**Pull request merged**